packages/ckeditor5-engine/src/dataprocessor/htmldataprocessor.js

```diff
@@ -119,10 +119,36 @@ export default class HtmlDataProcessor {
 	_toDom( data ) {
 		const document = this._domParser.parseFromString( data, 'text/html' );
 		const fragment = document.createDocumentFragment();
-		const nodes = document.body.childNodes;
 
-		while ( nodes.length > 0 ) {
-			fragment.appendChild( nodes[ 0 ] );
+		// The rules for parsing an HTML string can be read on https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inhtml.
+		//
+		// In short, parsing tokens in an HTML string starts with the so-called "initial" insertion mode. When a DOM parser is in this state
+		// and encounters a comment node, it inserts this comment node as the last child of the newly-created Document object. The parser
+		// then proceeds to successive insertion modes during parsing subsequent tokens and appends in the Document object other nodes (like
+		// <html>, <head>, <body>). This causes that the first leading comments from HTML string become the first nodes in the Document
+		// object, but not in the <body> collection, because they are ultimately located before the <html> element.
+		//
+		// Therefore, so that such leading comments do not disappear, they all are moved from the Document object to the document fragment,
+		// until the <html> element is encountered.
+		//
+		// See: https://github.com/ckeditor/ckeditor5/issues/9861.
+		let documentChildNode = document.firstChild;
+
+		while ( !documentChildNode.isSameNode( document.documentElement ) ) {
+			const node = documentChildNode;
+
+			documentChildNode = documentChildNode.nextSibling;
+
+			/* istanbul ignore else */
+			if ( node.nodeType == Node.COMMENT_NODE ) {
+				fragment.appendChild( node );
+			}
```

Comment:
Probably only comment nodes can be added directly to the Document object, before the `<html>` node. I have not been able to create any HTML string that would insert a node other than a comment at the beginning of the Document object, so the `/* istanbul ignore else */` is here to keep the 100% CC.

Anyway, this `if` condition here is just to be sure.

Reply:
Maybe it could be moved to the comment.

Reply:
Comment has been added.

I also replaced the `Document` with `HTMLDocument` in comments, as the latter name [is used in the specification](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser/parseFromString#return_value).